mass-deployer/pkg/mass-deployer/deployer.go

```diff
@@ -0,0 +1,283 @@
+package deployer
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"gopkg.in/yaml.v3"
+
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/tfgrid-sdk-go/grid-client/deployer"
+	"github.com/threefoldtech/tfgrid-sdk-go/grid-client/workloads"
+	"github.com/threefoldtech/tfgrid-sdk-go/grid-proxy/pkg/types"
+	"github.com/threefoldtech/zos/pkg/gridtypes"
+)
+
+type vmInfo struct {
+	Name      string
+	PublicIP4 string
+	PublicIP6 string
+	YggIP     string
+	IP        string
+	Mounts    []workloads.Mount
+}
+
+type groupDeploymentsInfo struct {
+	vmDeployments      []*workloads.Deployment
+	networkDeployments []*workloads.ZNet
+	deploymentsInfo    []vmDeploymentInfo
+}
+
+type vmDeploymentInfo struct {
+	nodeID         uint32
+	vmName         string
+	deploymentName string
+}
+
+func RunDeployer(cfg Config) error {
+	ctx := context.Background()
+	passedGroups := map[string][]string{}
+	failedGroups := map[string]error{}
+
+	tfPluginClient, err := setup(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create deployer: %v", err)
+	}
+
+	groupsNodes, failed := filterNodes(tfPluginClient, cfg.NodeGroups, ctx)
+	failedGroups = failed
+
+	groupsDeployments := parseVMs(tfPluginClient, cfg.Vms, groupsNodes, cfg.SSHKeys)
+	var lock sync.Mutex
+	var wg sync.WaitGroup
+
+	deploymentStart := time.Now()
+
+	for nodeGroup, deployemnts := range groupsDeployments {
+		wg.Add(1)
+		go func(group string, deployemnts groupDeploymentsInfo) {
+			defer wg.Done()
+			info, err := massDeploy(tfPluginClient, ctx, deployemnts)
```

Comment:
also if I remember correctly, I think there is an issue with deploying concurrently as multiple requests to the chain with the same mnemonic is not possible but I am not sure.